HW0/ans/P5.txt

```diff
@@ -0,0 +1,11 @@
+1. 255 seconds.
+2. Yes, it's faster.
+Each cashier counts 32 bags, which costs 31 seconds. Then they need to count these 8 sums, which costs 3 seconds. In total, in 34 seconds.
+3. 8 seconds.
+4. Using one cashier, it can be verified in N-1 seconds.
+Using infinite cashiers, it can be verified in [log2(N)]+1 seconds.
+5. Using 128 cashiers, each person counts 2 bags at first, costing 1 second. Then they count these 8 sums, costing 7*2=14 seconds. In total, it costs 15 seconds in total.
+6. (1) If 1 bag each time, it takes 260 seconds. 
+   (2) If two bags each time, it takes 133 seconds.
+   (3) It's better to take 16 bags each person, which costs 35 seconds in total.
```

Comment:
Giving your two perspectives to consider.
My strategy was to give decreasing number of bags to cashiers at the back of the queue, so that the cashiers at the front can start summing bags while you are handling out bags.
1st second: 22 bags to cashier 1
2nd second: 21 bags to cashier 2, cashier 1 starts counting (it will take him 21 seconds)
3rd second: 20 bags to cashier 3, cashier 2 starts counting (it will take him 20 seconds)
This method will take 33 seconds to finish (2 seconds more optimal)

The solution that I got as a comment is much better:
1st second: 256 bags to cashier 1
2nd second: cashier 1 gives 128 bags to cashier 2
3rd second: cashier 1 and cashier 2 give 64 bags to cashier 3 and cashier 4
4th second: cashier 1, 2, 3, 4 give 32 bags to cashier 5, 6, 7, 8
...

Essentially you are outsourcing the bags distributions to the cashiers, who can distribute the bags out much more efficiently than you. This method takes only 22 seconds!